File: thorp_toolkit/python/thorp_toolkit/progress_tracker.py
from .geometry import distance_2d
from .visualization import Visualization


class ProgressTracker:
    """ Tracks robot progression along a sequence of waypoints """

    def __init__(self, waypoints, reached_threshold):
        if type(waypoints) != list or not waypoints:
            raise Exception("Waypoints must be a non-empty list of poses")
        self._waypoints = waypoints
        self._reached_threshold = reached_threshold
        self._next_wp = 0
        self._reached = False
        self._min_dist = float('inf')
        # visualize semi-transparent waypoints; will become solid once reached
        for wp in self._waypoints:
            Visualization().add_disc_marker(wp, (0.2, 0.2, 0.000001), (0, 0, 1.0, 0.25))
        Visualization().publish_markers()

    def reset(self):
        self._next_wp = 0
        self._reached = False
        self._min_dist = float('inf')
        Visualization().delete_markers()

    def update_pose(self, robot_pose):
        """
        Update progress with a new robot pose.
        :param robot_pose: Current robot pose (stamped)
        """
        if self._next_wp is None:
            return  # already arrived
        dist = distance_2d(robot_pose, self._waypoints[self._next_wp])
        if self._reached and dist > self._min_dist:
            Visualization().add_disc_marker(self._waypoints[self._next_wp], (0.2, 0.2, 0.000001))
            Visualization().publish_markers()
            self._next_wp += 1
            if self._next_wp < len(self._waypoints):
                # go for next waypoint
                self._reached = False
                self._min_dist = float('inf')
            else:
                # arrived; clear reached waypoints markers
                Visualization().delete_markers()
                self._next_wp = None
            return

        self._min_dist = min(self._min_dist, dist)
        if dist <= self._reached_threshold:
            self._reached = True

    def next_waypoint(self):
        return self._next_wp

    def reached_waypoint(self):
        if self._next_wp is None:
            return len(self._waypoints) - 1
        elif self._next_wp == 0:
            return None
        return self._next_wp - 1

File: thorp_toolkit/src/progress_tracker.cpp
#include "thorp_toolkit/progress_tracker.hpp"

#include <thorp_toolkit/geometry.hpp>

namespace thorp::toolkit
{

ProgressTracker::ProgressTracker() : viz_("waypoints")
{
}

void ProgressTracker::init(const std::vector<geometry_msgs::PoseStamped>& waypoints, double reached_threshold)
{
  if (waypoints.empty())
  {
    throw std::invalid_argument("Waypoints must be a non-empty list of poses");
  }

  next_wp_ = 0;
  reached_ = false;
  min_dist_ = std::numeric_limits<double>::infinity();
  waypoints_ = waypoints;
  reached_threshold_ = reached_threshold;

  // Visualize semi-transparent waypoints; will become solid once reached_
  for (const auto& wp : waypoints_)
  {
    viz_.addDiscMarker(wp, 0.2, makeColor(0, 0, 1.0, 0.25));
  }
  viz_.publishMarkers();
}

void ProgressTracker::reset()
{
  waypoints_.clear();
  next_wp_ = 0;
  reached_ = false;
  min_dist_ = std::numeric_limits<double>::infinity();
  viz_.reset();
}

void ProgressTracker::updatePose(const geometry_msgs::PoseStamped& robot_pose)
{
  if (waypoints_.empty())
  {
    throw std::invalid_argument("Not initialized");
  }

  if (next_wp_ == std::numeric_limits<size_t>::max())
  {
    return;  // already arrived
  }

  double dist = distance2D(robot_pose, waypoints_[next_wp_]);
  if (reached_ && dist > min_dist_)
  {
    viz_.addDiscMarker(waypoints_[next_wp_], 0.2, makeColor(0, 0, 1.0, 1.0));
    viz_.publishMarkers();
    next_wp_ += 1;

    if (next_wp_ < waypoints_.size())
    {
      // go for the next waypoint
      reached_ = false;
      min_dist_ = std::numeric_limits<double>::infinity();
    }
    else
    {
      // arrived; clear reached waypoints markers
      viz_.deleteMarkers();
      next_wp_ = std::numeric_limits<size_t>::max();
    }
    return;
  }

  min_dist_ = std::min(min_dist_, dist);
  if (dist <= reached_threshold_)
  {
    reached_ = true;
  }
}

size_t ProgressTracker::nextWaypoint() const
{
  if (waypoints_.empty())
  {
    throw std::invalid_argument("Not initialized");
  }

  return next_wp_;
}

size_t ProgressTracker::reachedWaypoint() const
{
  if (waypoints_.empty())
  {
    throw std::invalid_argument("Not initialized");
  }

  if (next_wp_ == std::numeric_limits<size_t>::max())
  {
    return waypoints_.size() - 1;
  }
  else if (next_wp_ == 0)
  {
    return std::numeric_limits<size_t>::max();
  }
  return next_wp_ - 1;
}

}  // namespace thorp::toolkit
